javaagent-bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/ClassLoaderMatcherCacheHolder.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.bootstrap;
+
+import io.opentelemetry.instrumentation.api.caching.Cache;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import org.checkerframework.checker.lock.qual.GuardedBy;
+
+/**
+ * A holder of all ClassLoaderMatcher caches. We store them in the bootstrap classloader so that
+ * instrumentation can invalidate the ClassLoaderMatcher for a particular ClassLoader, e.g. when
+ * {@link java.net.URLClassLoader#addURL(URL)} is called.
+ */
+public class ClassLoaderMatcherCacheHolder {
+
+  @GuardedBy("allCaches")
+  private static final List<Cache<ClassLoader, Boolean>> allCaches = new ArrayList<>();
+
+  private ClassLoaderMatcherCacheHolder() {}
+
+  public static void addCache(Cache<ClassLoader, Boolean> cache) {
+    allCaches.add(cache);
+  }
+
+  public static void invalidateAllCachesForClassLoader(ClassLoader loader) {
+    List<Cache<ClassLoader, Boolean>> allCachesCopy;
+    synchronized (allCaches) {
```

Comment:
If we synchronize on `add` (which we should), I don't think we need this copy, can just synchronize this method.

Reply:
I get nervous calling anything under lock that I don't know exactly what it does (in this case `cache.remove()`), u know that code well though, so I'm good taking your recommendation

Reply:
Do you mean for performance reasons? Yeah I wouldn't worry about it for removing from a cache. Without the synchronized around the method, this code is harder to reason about - I think it actually is thread-safe only because we know `addURL` and `loadClass` are guarded by the same lock. If they weren't ,then a `loadClass` adding a new cache object could happen between copy and invalidate.

Reply:
my nervousness is about deadlocks 😁

> `loadClass` adding a new cache object could happen between copy and invalidate

oh, I hadn't thought about this, done 👍